File: android/accessibility-service/src/main/kotlin/dev/jasonpearson/automobile/accessibilityservice/HierarchyDebouncer.kt
package dev.jasonpearson.automobile.accessibilityservice

import android.util.Log
import dev.jasonpearson.automobile.accessibilityservice.models.ViewHierarchy
import dev.jasonpearson.automobile.accessibilityservice.perf.PerfProvider
import dev.jasonpearson.automobile.accessibilityservice.perf.SystemTimeProvider
import dev.jasonpearson.automobile.accessibilityservice.perf.TimeProvider
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.launch

/** Result of hierarchy extraction with hash comparison. */
sealed class HierarchyResult {
  /** New hierarchy extracted with different structural content. */
  data class Changed(val hierarchy: ViewHierarchy, val hash: Int, val extractionTimeMs: Long) :
      HierarchyResult()

  /**
   * Hierarchy extracted but structure unchanged (animation only). Bounds may have changed but
   * text/content is the same.
   */
  data class Unchanged(
      val hierarchy: ViewHierarchy,
      val hash: Int,
      val extractionTimeMs: Long,
      val skippedEventCount: Int,
  ) : HierarchyResult()

  /** Failed to extract hierarchy. */
  data class Error(val message: String) : HierarchyResult()
}

/**
 * Smart debouncer for view hierarchy extraction.
 *
 * Uses structural hash comparison to detect when content actually changes vs. when only bounds are
 * changing during animations.
 *
 * Key optimization: During scroll/fling animations, many TYPE_WINDOW_CONTENT_CHANGED events fire
 * but only bounds change. By comparing structural hashes, we can:
 * - Detect animation mode: Same hash = skip waiting, broadcast immediately
 * - Detect real changes: Different hash = content changed, proceed normally
 *
 * This reduces observation latency from 600ms+ to ~50ms during animations.
 *
 * @param scope Coroutine scope for debounce jobs
 * @param timeProvider Time provider for testability
 * @param perfProvider Performance tracking provider
 * @param quickDebounceMs Short debounce to batch rapid events (default 5ms)
 * @param animationSkipWindowMs How long to skip extractions after detecting animation (default
 *   100ms)
 * @param extractHierarchy Function to extract the current hierarchy
 */
class HierarchyDebouncer(
    private val scope: CoroutineScope,
    private val timeProvider: TimeProvider = SystemTimeProvider(),
    private val perfProvider: PerfProvider = PerfProvider.instance,
    private val quickDebounceMs: Long = 5L,
    private val animationSkipWindowMs: Long = 100L,
    private val extractHierarchy: (disableAllFiltering: Boolean) -> ViewHierarchy?,
) {
  companion object {
    private const val TAG = "HierarchyDebouncer"
  }

  // Last known structural hash
  private var lastStructuralHash: Int = 0

  // Animation mode tracking
  private var inAnimationMode: Boolean = false
  private var animationModeEndTime: Long = 0
  private var skippedEventCount: Int = 0

  // Debounce job
  private var debounceJob: Job? = null

  // Flag to temporarily suppress flow emissions (during setText operations)
  @Volatile private var suppressFlowEmissions: Boolean = false

  // Timestamp of last accessibility event (for quiescence detection)
  @Volatile private var lastEventTimestamp: Long = 0

  // Flow for emitting hierarchy results
  private val _hierarchyFlow =
      MutableSharedFlow<HierarchyResult>(replay = 1, extraBufferCapacity = 10)
  val hierarchyFlow: SharedFlow<HierarchyResult> = _hierarchyFlow

  // Last emitted hierarchy (for quick access)
  @Volatile private var lastHierarchy: ViewHierarchy? = null

  /**
   * Called when an accessibility event is received that might change the hierarchy. Handles
   * debouncing and decides whether to extract based on animation detection.
   */
  fun onAccessibilityEvent() {
    val now = timeProvider.currentTimeMillis()
    lastEventTimestamp = now

    // Skip all extractions if flow emissions are suppressed (setText/imeAction in progress)
    // The operation will call extractAfterQuiescence which does its own final extraction
    if (suppressFlowEmissions) {
      Log.d(TAG, "Skipping extraction (operation in progress, flow suppressed)")
      return
    }

    // If we're in animation mode and within the skip window, skip extraction
    if (inAnimationMode && now < animationModeEndTime) {
      skippedEventCount++
      Log.d(TAG, "Skipping extraction (animation mode), skipped: $skippedEventCount")
      return
    }

    // Exit animation mode if window expired
    if (inAnimationMode && now >= animationModeEndTime) {
      Log.d(TAG, "Exiting animation mode after skipping $skippedEventCount events")
      inAnimationMode = false
    }

    // Cancel previous debounce and start new one
    debounceJob?.cancel()
    debounceJob =
        scope.launch {
          delay(quickDebounceMs)
          extractAndCompare()
        }
  }

  /**
   * Perform an immediate extraction (bypasses debounce and animation mode). This is async - returns
   * immediately after launching the extraction.
   */
  fun extractNow(disableAllFiltering: Boolean = false) {
    debounceJob?.cancel()
    inAnimationMode = false
    scope.launch { extractAndCompare(disableAllFiltering = disableAllFiltering) }
  }

  /**
   * Perform an immediate extraction and wait for it to complete (blocking). This ensures the
   * hierarchy is extracted and pushed before returning. Use this when you need to guarantee the
   * hierarchy is fresh before sending a result.
   *
   * @param skipFlowEmit If true, skips emitting to the flow. Use this when the caller will
   *   broadcast the hierarchy directly to avoid race conditions with flow-based async broadcasts.
   * @param disableAllFiltering If true, disables filtering/optimization for this extraction.
   * @return The extracted hierarchy, or null if extraction failed.
   */
  fun extractNowBlocking(
      skipFlowEmit: Boolean = false,
      disableAllFiltering: Boolean = false,
  ): ViewHierarchy? {
    debounceJob?.cancel()
    inAnimationMode = false
    kotlinx.coroutines.runBlocking {
      extractAndCompare(skipFlowEmit = skipFlowEmit, disableAllFiltering = disableAllFiltering)
    }
    return lastHierarchy
  }

  /**
   * Wait for accessibility events to stop firing (quiescence), then extract hierarchy. This is the
   * preferred method for operations that modify UI state (like setText).
   *
   * Instead of using a fixed delay, this method dynamically waits until no accessibility events
   * have been received for [quiescenceMs], then extracts the hierarchy.
   *
   * Flow emissions are suppressed during this operation to prevent race conditions where debounced
   * extractions could broadcast stale hierarchies.
   *
   * @param quiescenceMs How long to wait with no events before considering UI settled (default
   *   50ms)
   * @param maxWaitMs Maximum time to wait for quiescence (default 500ms)
   * @param pollIntervalMs How often to check for quiescence (default 10ms)
   * @return The extracted hierarchy, or null if extraction failed.
   */
  fun extractAfterQuiescence(
      quiescenceMs: Long = 50L,
      maxWaitMs: Long = 500L,
      pollIntervalMs: Long = 10L,
      initialEventWaitMs: Long = 200L, // Max time to wait for first event
  ): ViewHierarchy? {
    val startTime = timeProvider.currentTimeMillis()
    val initialTimestamp = lastEventTimestamp // Capture timestamp BEFORE the action

    // Suppress flow emissions to prevent racing broadcasts from debounced extractions
    suppressFlowEmissions = true
    Log.d(
        TAG,
        "extractAfterQuiescence: suppressing flow emissions, waiting for first event then quiescence (${quiescenceMs}ms quiet)",
    )

    try {
      // Cancel any pending debounce job
      debounceJob?.cancel()
      inAnimationMode = false

      kotlinx.coroutines.runBlocking {
        // PHASE 1: Wait for at least one new accessibility event to fire
        // This is critical because ACTION_SET_TEXT completes before the accessibility
        // tree is updated. The tree update triggers TYPE_WINDOW_CONTENT_CHANGED events.
        var sawFirstEvent = false
        while (!sawFirstEvent) {
          val now = timeProvider.currentTimeMillis()
          val elapsed = now - startTime

          // Check if we've exceeded initial event wait time
          if (elapsed >= initialEventWaitMs) {
            Log.d(TAG, "extractAfterQuiescence: no events after ${elapsed}ms, proceeding anyway")
            break
          }

          // Check if a new event has fired since we started
          if (lastEventTimestamp != initialTimestamp) {
            Log.d(TAG, "extractAfterQuiescence: first event detected after ${elapsed}ms")
            sawFirstEvent = true
            break
          }

          delay(pollIntervalMs)
        }

        // PHASE 2: Wait for quiescence (no events for quiescenceMs)
        var lastCheckedTimestamp = lastEventTimestamp

        while (true) {
          val now = timeProvider.currentTimeMillis()
          val elapsed = now - startTime
          val timeSinceLastEvent = now - lastEventTimestamp

          // Check if we've exceeded max wait time
          if (elapsed >= maxWaitMs) {
            Log.d(
                TAG,
                "extractAfterQuiescence: max wait time exceeded (${elapsed}ms), proceeding with extraction",
            )
            break
          }

          // Check if we've achieved quiescence
          if (timeSinceLastEvent >= quiescenceMs) {
            Log.d(
                TAG,
                "extractAfterQuiescence: quiescence achieved after ${elapsed}ms (no events for ${timeSinceLastEvent}ms)",
            )
            break
          }

          // Log if new event was detected
          if (lastEventTimestamp != lastCheckedTimestamp) {
            Log.d(TAG, "extractAfterQuiescence: new event detected, resetting quiescence timer")
            lastCheckedTimestamp = lastEventTimestamp
          }

          // Wait before checking again
          delay(pollIntervalMs)
        }

        // Now extract the hierarchy
        extractAndCompare(skipFlowEmit = true)
      }

      val totalWait = timeProvider.currentTimeMillis() - startTime
      Log.d(TAG, "extractAfterQuiescence: completed in ${totalWait}ms")

      return lastHierarchy
    } finally {
      // Always unsuppress flow emissions
      suppressFlowEmissions = false
      Log.d(TAG, "extractAfterQuiescence: flow emissions re-enabled")
    }
  }

  /**
   * Extract hierarchy and compare structural hash.
   *
   * IMPORTANT: The perf block MUST be closed BEFORE calling _hierarchyFlow.emit() because emit()
   * can suspend waiting for collectors. If a new accessibility event triggers another
   * extractAndCompare() while we're suspended on emit, its perf block would become a child of this
   * still-open block, causing nested accumulation.
   *
   * @param skipFlowEmit If true, skips emitting to the flow. Use this when the caller will
   *   broadcast directly to avoid race conditions.
   */
  private suspend fun extractAndCompare(
      skipFlowEmit: Boolean = false,
      disableAllFiltering: Boolean = false,
  ) {
    val startTime = timeProvider.currentTimeMillis()

    // Variables to hold results for emission after perf block closes
    var resultToEmit: HierarchyResult? = null
    var hierarchyToCache: ViewHierarchy? = null

    // Use independentRoot so concurrent extractions are tracked as parallel siblings
    // rather than nested within each other
    perfProvider.independentRoot("hierarchyDebouncer")
    try {
      perfProvider.startOperation("extractHierarchy")
      val hierarchy = extractHierarchy(disableAllFiltering)
      perfProvider.endOperation("extractHierarchy")

      if (hierarchy == null) {
        resultToEmit = HierarchyResult.Error("Failed to extract hierarchy")
        return
      }

      val extractionTime = timeProvider.currentTimeMillis() - startTime

      perfProvider.startOperation("computeHash")
      val structuralHash = StructuralHasher.computeHash(hierarchy)
      perfProvider.endOperation("computeHash")

      if (structuralHash == lastStructuralHash) {
        // Structure unchanged - likely animation
        // Enter animation mode to skip subsequent events
        inAnimationMode = true
        animationModeEndTime = timeProvider.currentTimeMillis() + animationSkipWindowMs

        resultToEmit =
            HierarchyResult.Unchanged(
                hierarchy = hierarchy,
                hash = structuralHash,
                extractionTimeMs = extractionTime,
                skippedEventCount = skippedEventCount,
            )
        hierarchyToCache = hierarchy

        Log.d(
            TAG,
            "Structure unchanged (hash=$structuralHash), entering animation mode for ${animationSkipWindowMs}ms",
        )

        // Reset skipped count
        skippedEventCount = 0
      } else {
        // Structure changed - this is a real content change
        val oldHash = lastStructuralHash
        inAnimationMode = false
        lastStructuralHash = structuralHash

        resultToEmit =
            HierarchyResult.Changed(
                hierarchy = hierarchy,
                hash = structuralHash,
                extractionTimeMs = extractionTime,
            )
        hierarchyToCache = hierarchy

        Log.d(TAG, "Structure changed (oldHash=$oldHash, newHash=$structuralHash)")

        skippedEventCount = 0
      }
    } finally {
      // End perf block BEFORE emit to prevent nesting if emit suspends
      perfProvider.end()
    }

    // Emit AFTER perf block is closed - this can suspend without causing nesting issues
    hierarchyToCache?.let { lastHierarchy = it }
    // Only emit to flow if not skipped AND emissions are not suppressed
    // suppressFlowEmissions is used during setText operations to prevent racing broadcasts
    if (!skipFlowEmit && !suppressFlowEmissions) {
      resultToEmit?.let { _hierarchyFlow.emit(it) }
    } else if (suppressFlowEmissions) {
      Log.d(TAG, "Flow emission suppressed (setText operation in progress)")
    }
  }

  /** Get the last extracted hierarchy without triggering a new extraction. */
  fun getLastHierarchy(): ViewHierarchy? = lastHierarchy

  /**
   * Get the timestamp of the last accessibility event. Used by the TypeScript MCP server to check
   * if events have occurred since a request was made.
   */
  fun getLastEventTimestamp(): Long = lastEventTimestamp

  /**
   * Check if any accessibility events have occurred since the given timestamp. If no events have
   * occurred (stale), trigger an immediate extraction. This is used when the MCP server is waiting
   * for pushed data but no events are firing.
   *
   * @param sinceTimestamp The timestamp to compare against (usually the MCP request start time)
   */
  fun extractIfStale(sinceTimestamp: Long) {
    if (lastEventTimestamp <= sinceTimestamp) {
      // No events since the given timestamp - the UI is "stale" from MCP's perspective
      // Trigger immediate extraction to push current state
      Log.d(
          TAG,
          "extractIfStale: no events since $sinceTimestamp (lastEvent=$lastEventTimestamp), triggering extraction",
      )
      extractNow()
    } else {
      Log.d(
          TAG,
          "extractIfStale: events occurred since $sinceTimestamp (lastEvent=$lastEventTimestamp), skipping extraction",
      )
    }
  }

  /** Get current state for debugging. */
  fun getState(): DebounceState {
    return DebounceState(
        lastHash = lastStructuralHash,
        inAnimationMode = inAnimationMode,
        animationModeEndTime = animationModeEndTime,
        skippedEventCount = skippedEventCount,
        hasActiveJob = debounceJob?.isActive == true,
    )
  }

  /** Reset all state (for testing or reconnection). */
  fun reset() {
    debounceJob?.cancel()
    debounceJob = null
    lastStructuralHash = 0
    inAnimationMode = false
    animationModeEndTime = 0
    skippedEventCount = 0
    lastHierarchy = null
  }

  /** Debugging state info. */
  data class DebounceState(
      val lastHash: Int,
      val inAnimationMode: Boolean,
      val animationModeEndTime: Long,
      val skippedEventCount: Int,
      val hasActiveJob: Boolean,
  )
}

File: ios/XCTestService/Sources/XCTestService/HierarchyDebouncer.swift
import Foundation

/// Result of hierarchy extraction with hash comparison.
public enum HierarchyResult {
    /// New hierarchy extracted with different structural content.
    case changed(hierarchy: ViewHierarchy, hash: Int, extractionTimeMs: Int64)

    /// Hierarchy extracted but structure unchanged (animation only).
    case unchanged(hierarchy: ViewHierarchy, hash: Int, extractionTimeMs: Int64, skippedPollCount: Int)

    /// Failed to extract hierarchy.
    case error(message: String)
}

/// Protocol for hierarchy debouncing
public protocol HierarchyDebouncing {
    /// Start polling for changes
    func start()

    /// Stop polling for changes
    func stop()

    /// Whether the debouncer is currently running
    var isRunning: Bool { get }

    /// Perform an immediate extraction (bypasses debounce and animation mode)
    func extractNow()

    /// Perform an immediate extraction and wait for it to complete (blocking)
    func extractNowBlocking(skipFlowEmit: Bool) -> ViewHierarchy?

    /// Set callback for hierarchy results
    func setOnResult(_ callback: @escaping (HierarchyResult) -> Void)

    /// Get the last extracted hierarchy without triggering a new extraction
    func getLastHierarchy() -> ViewHierarchy?

    /// Reset all state
    func reset()
}

/// Smart debouncer for view hierarchy extraction on iOS.
///
/// Uses structural hash comparison to detect when content actually changes vs. when only
/// bounds are changing during animations.
///
/// Key optimization: During animations, many UI updates fire but only bounds change.
/// By comparing structural hashes, we can:
/// - Detect animation mode: Same hash = skip broadcasting, continue polling
/// - Detect real changes: Different hash = content changed, broadcast
///
/// This reduces noise during animations while still detecting real changes quickly.
public class HierarchyDebouncer: HierarchyDebouncing {
    // MARK: - Configuration

    /// How often to poll for changes (default 10ms for quick detection)
    public static let defaultPollIntervalMs: Int64 = 10

    /// How long to skip broadcasts after detecting animation (default 100ms)
    public static let animationSkipWindowMs: Int64 = 100

    /// Minimum interval between broadcasts (debounce)
    public static let broadcastDebounceMs: Int64 = 50

    // MARK: - Dependencies

    private let elementLocator: ElementLocating
    private let timer: Timer
    private let pollIntervalMs: Int64

    // MARK: - State

    private var lastStructuralHash = 0
    private var inAnimationMode = false
    private var animationModeEndTime: Int64 = 0
    private var skippedPollCount = 0
    private var lastBroadcastTime: Int64 = 0
    private var lastHierarchy: ViewHierarchy?

    private let lock = NSLock()
    private var _isRunning = false
    private var pollScheduled = false
    private var onResult: ((HierarchyResult) -> Void)?

    public var isRunning: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _isRunning
    }

    // MARK: - Init

    public init(
        elementLocator: ElementLocating,
        timer: Timer = SystemTimer(),
        pollIntervalMs: Int64 = HierarchyDebouncer.defaultPollIntervalMs
    ) {
        self.elementLocator = elementLocator
        self.timer = timer
        self.pollIntervalMs = pollIntervalMs
    }

    // MARK: - Public Interface

    public func setOnResult(_ callback: @escaping (HierarchyResult) -> Void) {
        lock.lock()
        defer { lock.unlock() }
        onResult = callback
    }

    public func start() {
        lock.lock()
        guard !_isRunning else {
            lock.unlock()
            return
        }
        _isRunning = true
        lock.unlock()

        // Capture initial state
        captureInitialState()

        // Schedule first poll
        scheduleNextPoll()

        print("[HierarchyDebouncer] Started with \(pollIntervalMs)ms polling interval")
    }

    public func stop() {
        lock.lock()
        _isRunning = false
        pollScheduled = false
        lock.unlock()

        print("[HierarchyDebouncer] Stopped")
    }

    public func extractNow() {
        lock.lock()
        inAnimationMode = false
        lock.unlock()

        extractAndCompare(skipBroadcast: false)
    }

    public func extractNowBlocking(skipFlowEmit: Bool = false) -> ViewHierarchy? {
        lock.lock()
        inAnimationMode = false
        lock.unlock()

        extractAndCompare(skipBroadcast: skipFlowEmit)

        lock.lock()
        let hierarchy = lastHierarchy
        lock.unlock()
        return hierarchy
    }

    public func getLastHierarchy() -> ViewHierarchy? {
        lock.lock()
        defer { lock.unlock() }
        return lastHierarchy
    }

    public func reset() {
        lock.lock()
        lastStructuralHash = 0
        inAnimationMode = false
        animationModeEndTime = 0
        skippedPollCount = 0
        lastBroadcastTime = 0
        lastHierarchy = nil
        lock.unlock()
    }

    // MARK: - Private

    private func scheduleNextPoll() {
        lock.lock()
        guard _isRunning, !pollScheduled else {
            lock.unlock()
            return
        }
        pollScheduled = true
        lock.unlock()

        timer.schedule(after: pollIntervalMs) { [weak self] in
            guard let self = self else { return }

            self.lock.lock()
            self.pollScheduled = false
            let shouldContinue = self._isRunning
            self.lock.unlock()

            if shouldContinue {
                self.pollAndCheck()
                self.scheduleNextPoll()
            }
        }
    }

    private func captureInitialState() {
        do {
            let hierarchy = try elementLocator.getViewHierarchy(disableAllFiltering: false)
            let hash = StructuralHasher.computeHash(hierarchy)

            lock.lock()
            lastStructuralHash = hash
            lastHierarchy = hierarchy
            lock.unlock()

            print("[HierarchyDebouncer] Initial state captured (hash=\(hash))")
        } catch {
            print("[HierarchyDebouncer] Failed to capture initial state: \(error)")
        }
    }

    private func pollAndCheck() {
        let now = timer.now()

        lock.lock()
        let running = _isRunning
        let animationMode = inAnimationMode
        let animationEnd = animationModeEndTime
        lock.unlock()

        guard running else { return }

        // If we're in animation mode and within the skip window, skip extraction
        if animationMode, now < animationEnd {
            lock.lock()
            skippedPollCount += 1
            lock.unlock()
            return
        }

        // Exit animation mode if window expired
        if animationMode, now >= animationEnd {
            lock.lock()
            let skipped = skippedPollCount
            inAnimationMode = false
            lock.unlock()
            print("[HierarchyDebouncer] Exiting animation mode after skipping \(skipped) polls")
        }

        extractAndCompare(skipBroadcast: false)
    }

    private func extractAndCompare(skipBroadcast: Bool) {
        let startTime = timer.now()

        do {
            let hierarchy = try elementLocator.getViewHierarchy(disableAllFiltering: false)
            let extractionTime = timer.now() - startTime
            let newHash = StructuralHasher.computeHash(hierarchy)

            lock.lock()
            let oldHash = lastStructuralHash
            let callback = onResult
            let lastBroadcast = lastBroadcastTime
            lock.unlock()

            if newHash == oldHash {
                // Structure unchanged - likely animation
                lock.lock()
                inAnimationMode = true
                animationModeEndTime = timer.now() + HierarchyDebouncer.animationSkipWindowMs
                lastHierarchy = hierarchy
                lock.unlock()

                // Don't broadcast unchanged results to reduce noise
                // Structure unchanged = animation mode, just reset counter
                lock.lock()
                skippedPollCount = 0
                lock.unlock()

            } else {
                // Structure changed - this is a real content change
                let now = timer.now()

                lock.lock()
                inAnimationMode = false
                lastStructuralHash = newHash
                lastHierarchy = hierarchy
                skippedPollCount = 0
                lock.unlock()

                // Debounce broadcasts
                let timeSinceLastBroadcast = now - lastBroadcast
                let shouldBroadcast = !skipBroadcast && timeSinceLastBroadcast >= HierarchyDebouncer.broadcastDebounceMs

                if shouldBroadcast {
                    lock.lock()
                    lastBroadcastTime = now
                    lock.unlock()

                    let result = HierarchyResult.changed(
                        hierarchy: hierarchy,
                        hash: newHash,
                        extractionTimeMs: extractionTime
                    )

                    print(
                        "[HierarchyDebouncer] Structure changed (oldHash=\(oldHash), newHash=\(newHash)), broadcasting"
                    )
                    callback?(result)
                }
            }
        } catch {
            // Silently ignore errors during polling
            // This can happen if the app is transitioning between states
        }
    }
}

// MARK: - Structural Hasher

/// Computes a structural hash of a ViewHierarchy for change detection.
/// Ignores bounds to focus on content changes vs. animation changes.
public enum StructuralHasher {
    /// Compute a structural hash of the hierarchy.
    /// Ignores bounds to differentiate content changes from animation/scroll changes.
    public static func computeHash(_ hierarchy: ViewHierarchy) -> Int {
        var hasher = Hasher()

        // Include package name
        if let packageName = hierarchy.packageName {
            hasher.combine(packageName)
        }

        // Include hierarchy structure (but not bounds)
        if let root = hierarchy.hierarchy {
            hashElement(root, into: &hasher, depth: 0, maxDepth: 15)
        }

        return hasher.finalize()
    }

    private static func hashElement(_ element: UIElementInfo, into hasher: inout Hasher, depth: Int, maxDepth: Int) {
        // Hash key identifying properties (NOT bounds - those change during animations)
        if let text = element.text {
            hasher.combine(text)
        }
        if let resourceId = element.resourceId {
            hasher.combine(resourceId)
        }
        if let className = element.className {
            hasher.combine(className)
        }

        // Hash state properties
        hasher.combine(element.focused)
        hasher.combine(element.selected)
        hasher.combine(element.checked)
        hasher.combine(element.enabled)

        // Hash children recursively (up to maxDepth)
        if depth < maxDepth, let children = element.node {
            hasher.combine(children.count)
            for child in children {
                hashElement(child, into: &hasher, depth: depth + 1, maxDepth: maxDepth)
            }
        }
    }
}

// MARK: - Fake for Testing

/// Fake implementation for testing hierarchy debouncing
public class FakeHierarchyDebouncer: HierarchyDebouncing {
    // Call tracking
    public private(set) var startCallCount = 0
    public private(set) var stopCallCount = 0
    public private(set) var extractNowCallCount = 0
    public private(set) var extractNowBlockingCallCount = 0
    public private(set) var setOnResultCallCount = 0
    public private(set) var resetCallCount = 0

    // State
    private var _isRunning = false
    private var onResult: ((HierarchyResult) -> Void)?
    private var lastHierarchy: ViewHierarchy?
    private let lock = NSLock()

    /// Configure what hierarchy to return from extractNowBlocking
    public var hierarchyToReturn: ViewHierarchy?

    public var isRunning: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _isRunning
    }

    public init() {}

    public func start() {
        lock.lock()
        startCallCount += 1
        _isRunning = true
        lock.unlock()
    }

    public func stop() {
        lock.lock()
        stopCallCount += 1
        _isRunning = false
        lock.unlock()
    }

    public func extractNow() {
        lock.lock()
        extractNowCallCount += 1
        lock.unlock()
    }

    public func extractNowBlocking(skipFlowEmit _: Bool = false) -> ViewHierarchy? {
        lock.lock()
        extractNowBlockingCallCount += 1
        let hierarchy = hierarchyToReturn
        lastHierarchy = hierarchy
        lock.unlock()
        return hierarchy
    }

    public func setOnResult(_ callback: @escaping (HierarchyResult) -> Void) {
        lock.lock()
        setOnResultCallCount += 1
        onResult = callback
        lock.unlock()
    }

    public func getLastHierarchy() -> ViewHierarchy? {
        lock.lock()
        defer { lock.unlock() }
        return lastHierarchy
    }

    public func reset() {
        lock.lock()
        resetCallCount += 1
        lastHierarchy = nil
        lock.unlock()
    }

    /// Simulate a hierarchy change for testing
    public func simulateChange(_ result: HierarchyResult) {
        lock.lock()
        let callback = onResult
        if case let .changed(hierarchy, _, _) = result {
            lastHierarchy = hierarchy
        } else if case let .unchanged(hierarchy, _, _, _) = result {
            lastHierarchy = hierarchy
        }
        lock.unlock()
        callback?(result)
    }

    /// Reset all call counts for fresh test assertions
    public func resetCounts() {
        lock.lock()
        startCallCount = 0
        stopCallCount = 0
        extractNowCallCount = 0
        extractNowBlockingCallCount = 0
        setOnResultCallCount = 0
        resetCallCount = 0
        lock.unlock()
    }
}
